File: pt5/1151.minimum-swaps-to-group-all-1-s-together.py
from __future__ import annotations

from typing import List


class Solution:
    def minSwaps(self, data: List[int]) -> int:
        ones = sum(data)
        if ones <= 1:
            return 0

        zeros = ones - sum(data[:ones])
        best = zeros
        for right in range(ones, len(data)):
            if data[right] == 0:
                zeros += 1
            if data[right - ones] == 0:
                zeros -= 1
            best = min(best, zeros)

        return best

File: pt5/1151.minimum-swaps-to-group-all-1-s-together.go
package main

func minSwaps(data []int) int {
	ones := 0
	for _, value := range data {
		ones += value
	}
	if ones <= 1 {
		return 0
	}

	zeros := 0
	for i := 0; i < ones; i++ {
		if data[i] == 0 {
			zeros++
		}
	}
	best := zeros
	for right := ones; right < len(data); right++ {
		if data[right] == 0 {
			zeros++
		}
		if data[right-ones] == 0 {
			zeros--
		}
		if zeros < best {
			best = zeros
		}
	}
	return best
}

